Ignore ~/.gitconfig when running git tests

ivanov opened the pull request:
Set HOME environment variable to the temporary directory the tests will
run in, avoiding any ~/.gitconfig settings that may interfere with
assumptions in the tests, particularly init.defaultBranch and
init.templateDir

Context: 
There are two settings that I'm aware of that can violate the assumption that the default branch will be called "master" when running `git init`. The older way is via the templateDir setting, which if pointed to a directory with a HEAD file containing `ref: refs/heads/main` will cause `git init` to default to `main` branch. Additionally, starting with version 2.28, git introduced a defaultBranch setting, that also allows developers to change the name of the initial branch name in a brand new repository.

This PR contains a minimal change to make the git-related tests pass locally (I use the templateDir method to change my default branch name).

Thank you for making and maintaining darker!

---

- [x] review #135 
- [x] merge #135 
- [x] rebase this one on `master`

akaihola commented:
Thanks @ivanov, this is a great catch!

I wonder if it would make sense to also run all the other Git commands besides `git init` this way as well, to prevent other possible side effects from exotic local Git configurations?

akaihola commented on src/darker/tests/conftest.py:
```diff
@@ -32,7 +33,10 @@ def __init__(self, root: Path):
-        check_call(["git", "init"], cwd=root)
+        env = os.environ.copy()
+        # for testsing, ignore ~/.gitconfig settings like templateDir and defaultBranch
+        env['HOME'] = root
```
```suggestion
        env["HOME"] = root
```

Darker's own test pipeline fails if non-Black-compliant code is committed on modified lines. And we don't the `--skip-string-normalization` option of Black anymore.

ivanov replied:
How embarrassing that I didn't run darker before submitting a pr to darker  :D 

fixed, thanks!

akaihola commented:
It would be good to have a unit test which fails without this patch. Also, it's nice to note the improvement in `README.rst`. If you're short on time, I can do those as well.

ivanov commented:
> I wonder if it would make sense to also run all the other Git commands besides git init this way as well, to prevent other possible side effects from exotic local Git configurations?

Done - now all tests for git commands use env where HOME points to the repo root.

> It would be good to have a unit test which fails without this patch. 

Done - here's what the a904a4305b991803da273fab100884c2fe9a82b7 cherry-picked to master looks like:

```
========================================= FAILURES ===================================
___________________ test_local_gitconfig_ignored_by_gitrepofixture ___________________
tmp_path = PosixPath('/tmp/pytest-of-pi/pytest-29/test_local_gitconfig_ignored_b0')

    def test_local_gitconfig_ignored_by_gitrepofixture(tmp_path):
        """Tests that ~/.gitconfig is ignored when running darker's git tests"""
        with open(tmp_path / "HEAD", "w") as head_file:
            head_file.write("ref: refs/heads/main")

        with patch.dict(os.environ, {"HOME": str(tmp_path)}):
            # Note: once we decide to drop support for git < 2.28, the HEAD file
            # creation above can be removed, and setup can simplify to
            # check_call("git config --global init.defaultBranch main".split())
            check_call("git config --global init.templateDir".split() + [tmp_path])
            root = tmp_path / "repo"
            root.mkdir()
            git_repo = GitRepoFixture.create_repository(root)
>           assert git_repo.get_branch() == "master"
E           AssertionError: assert 'main' == 'master'
E             - master
E             + main

src/darker/tests/test_git.py:467: AssertionError
```

> Also, it's nice to note the improvement in README.rst

I wasn't sure what section to put this in, but hopefully where I landed this was OK. Please feel free to revise or move it elsewhere as you see fit :)

akaihola commented on README.rst:
```diff
@@ -276,6 +276,10 @@ For example:
+*New in version 1.3.0:*
+
+- git-related commands in the test suite now ignore the user's ~/.gitconfig
+
```
Could you move this to `CHANGES.rst`?

akaihola commented on src/darker/tests/test_git.py:
```diff
@@ -449,3 +449,19 @@ def test_edited_linenums_differ_revision_vs_lines(git_repo, context_lines, expec
+
+def test_local_gitconfig_ignored_by_gitrepofixture(tmp_path):
+    """Tests that ~/.gitconfig is ignored when running darker's git tests"""
+    with open(tmp_path / "HEAD", "w") as head_file:
+        head_file.write("ref: refs/heads/main")
```
This could be written as
```suggestion
    (tmp_path / "HEAD").write_text("ref: refs/heads/main")
```

ivanov replied:
TIL. old habits die hard.

akaihola commented on src/darker/tests/test_main.py:
```diff
@@ -14,12 +13,13 @@
 
 
 def test_isort_option_without_isort(tmpdir, without_isort, caplog):
-    check_call(["git", "init"], cwd=tmpdir)
+    GitRepoFixture.create_repository(tmpdir)
```
Why not just use the `git_repo` fixture directly? It does exactly this (plus chdirs into the repo root).
```python
def test_isort_option_without_isort(git_repo, without_isort, caplog):
    # check_call(["git", "init"], cwd=tmpdir)  # remove this...
    # GitRepoFixture.create_repository(tmpdir)  # ...as well as this
```
The `GitRepoFixture` import should then be removed as well.

ivanov replied:
sure, ended up having to add a docstring to the test and `# pylint: disable=unused-argument` since neither `git_repo` nor `without_isort` end up getting used.

akaihola reviewed (changes requested):
Sorry for the delay @ivanov! Here are a couple of comments still. Let me know if you're busy and would like me to apply these tweaks.

akaihola commented:
Build fails due to Black having updated to 21.x. See #135 for a fix. Let's merge that one first and rebase.

akaihola commented on src/darker/tests/conftest.py:
```diff
@@ -26,20 +27,25 @@ def with_isort():
+        # for testing, ignore ~/.gitconfig settings like templateDir and defaultBranch
+        env["HOME"] = str(root)
+        instance = cls(root, env)
+        instance._run("init")
```
PyLint complains and fails our build unless we explicitly allow access to a protected method of another class instance:
```suggestion
        # pylint: disable=protected-access
        instance._run("init")
```

ivanov replied:
I added this, too, but for what it's worth, didn't see a complaint when running the test suite locally or via `darker -L pylint`  with black version 21.5b0 on Python 3.7.3.

akaihola replied:
Strange, I did see the complaint. Maybe we have different versions of Pylint.

ivanov commented:
Addressed the review comments, hopefully we're all good here, @akaihola , so back over to you.

akaihola commented on src/darker/tests/test_git.py:
```diff
@@ -449,3 +449,18 @@ def test_edited_linenums_differ_revision_vs_lines(git_repo, context_lines, expec
+        # Note: once we decide to drop support for git < 2.28, the HEAD file
+        # creation above can be removed, and setup can simplify to
+        # check_call("git config --global init.defaultBranch main".split())
+        check_call("git config --global init.templateDir".split() + [tmp_path])
```
It seems this needs to be turned into a string – py36 and py37 Windows builds are failing.
```suggestion
        check_call("git config --global init.templateDir".split() + [str(tmp_path)])
```

ivanov replied:
Thanks for this!

akaihola reviewed (changes requested):
A Windows build failure needs to be addressed.

akaihola commented:
Thanks @ivanov for this patch!
